rct_image_tools/src/rct_image_tools/modified_circle_grid_finder.cpp

```diff
@@ -106,38 +101,94 @@ static std::vector<cv::Point2d> extractKeyPoints(const cv::Mat& image, const std
+  double end_first_row_avg_rel_size = 0;
+  double end_last_row_avg_rel_size = 0;
+
   cv::Point2d start_last_row_pt = centers[start_last_row];
   cv::Point2d end_last_row_pt = centers[end_last_row];
   cv::Point2d start_first_row_pt = centers[start_first_row];
   cv::Point2d end_first_row_pt = centers[end_first_row];
 
+  // Get the size of every circle
+  std::vector<double> center_sizes(centers.size(), 1.0);
+
   for (std::size_t i = 0; i < keypoints.size(); i++)
   {
     double x = keypoints[i].pt.x;
     double y = keypoints[i].pt.y;
     double ksize = keypoints[i].size;
 
+    for (std::size_t j = 0; j < centers.size(); j++)
+    {
+      auto center = centers[j];
+      if (center.x == x && center.y == y)
+      {
+        center_sizes[j] = ksize;
+        break;
+      }
+    }
+  }
+
+  // Iterate over every center, when it is a corner then compare its size to adjacent circles
+  for (std::size_t i = 0; i < centers.size(); i++)
+  {
+    double x = centers[i].x;
+    double y = centers[i].y;
+    double ksize = center_sizes[i];
+
     if (x == start_last_row_pt.x && y == start_last_row_pt.y)
     {
       start_last_row_size = ksize;
+      if (i + 1 < center_sizes.size() && i - cols < center_sizes.size())
+      {
+        double start_last_row_rel_row_size = start_last_row_size / center_sizes[i+1];
+        double start_last_row_rel_col_size = start_last_row_size / center_sizes[i-cols];
+        start_last_row_avg_rel_size = (start_last_row_rel_row_size + start_last_row_rel_col_size)/2.0;
+      }
     }
     if (x == end_last_row_pt.x && y == end_last_row_pt.y)
     {
       end_last_row_size = ksize;
+      if (i - 1 < center_sizes.size() && i - cols < center_sizes.size())
+      {
+        double end_last_row_rel_row_size = end_last_row_size / center_sizes[i-1];
+        double end_last_row_rel_col_size = end_last_row_size / center_sizes[i-cols];
+        end_last_row_avg_rel_size = (end_last_row_rel_row_size + end_last_row_rel_col_size)/2.0;
+      }
     }
     if (x == start_first_row_pt.x && y == start_first_row_pt.y)
     {
       start_first_row_size = ksize;
+      if (i + 1 < center_sizes.size() && i + cols < center_sizes.size())
+      {
+        double start_first_row_rel_row_size = start_first_row_size / center_sizes[i+1];
+        double start_first_row_rel_col_size = start_first_row_size / center_sizes[i+cols];
+        start_first_row_avg_rel_size = (start_first_row_rel_row_size + start_first_row_rel_col_size)/2.0;
+      }
     }
     if (x == end_first_row_pt.x && y == end_first_row_pt.y)
     {
       end_first_row_size = ksize;
+      if (i - 1 < center_sizes.size() && i + cols < center_sizes.size())
+      {
+        double end_first_row_rel_row_size = end_first_row_size / center_sizes[i-1];
+        double end_first_row_rel_col_size = end_first_row_size / center_sizes[i+cols];
+        end_first_row_avg_rel_size = (end_first_row_rel_row_size + end_first_row_rel_col_size)/2.0;
+      }
```

Comment:
Please add some comments about what exactly you are comparing here to get the average relative size. I don't exactly follow which centers you are picking in each case for comparison